jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/TransformingFlusher.java

```diff
@@ -0,0 +1,207 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.websocket.core.Frame;
+
+/**
+ * This is used to iteratively transform or process a frame.
+ */
+public abstract class TransformingFlusher
+{
+    private final Logger log = Log.getLogger(this.getClass());
+
+    private final Queue<FrameEntry> entries = new ArrayDeque<>();
+    private final IteratingCallback flusher = new Flusher();
+    private boolean finished = true;
+    private Throwable failure;
+
+    /**
+     * Called when a frame is ready to be transformed.
+     * @param frame the frame to transform.
+     * @param callback used to signal to start processing again.
+     * @param batch whether this frame can be batched.
+     */
+    protected abstract void onFrame(Frame frame, Callback callback, boolean batch);
+
+    /**
+     * Called multiple times to transform the frame given in {@link TransformingFlusher#onFrame(Frame, Callback, boolean)}.
```

Comment:
possibly multiple times.... or possible not at all if setFinished(true) is called from onFrame